src/unix/syscall.c

```diff
@@ -1591,24 +1591,29 @@ static sysreturn brk(void *x)
     kernel_heaps kh = get_kernel_heaps();
 
     if (x) {
+        u64 old_end = pad(u64_from_pointer(p->brk), PAGESIZE);
+        u64 new_end = pad(u64_from_pointer(x), PAGESIZE);
         if (p->brk > x) {
             /* on failure, return the current break */
             if (u64_from_pointer(x) < p->heap_base)
                 goto fail;
             p->brk = x;
             assert(adjust_process_heap(p, irange(p->heap_base, u64_from_pointer(x))));
-            // free
+            if (old_end > new_end) {
+                u64 len = old_end - new_end;
+                u64 phys = physical_from_virtual(pointer_from_u64(new_end));
+                write_barrier();
+                unmap(new_end, len);
+                id_heap_set_area(heap_physical(kh), phys, len, true, false);
+            }
         } else if (p->brk < x) {
-            // I guess assuming we're aligned
-            u64 alloc = pad(u64_from_pointer(x), PAGESIZE) - pad(u64_from_pointer(p->brk), PAGESIZE);
+            u64 alloc = new_end - old_end;
+            assert(alloc > 0);
```

Comment:
Userspace should be allowed to call brk() with a non-page-aligned address, and the kernel should handle page alignment transparently for the program (meaning brk() should return the address requested by the program even if unaligned).

Reply:
Yes, I agree. I have further revised the brk(2) implementation and opened PR #1527 for it.